### PR TITLE
feat(routes): adds gatekeeper service

### PR DESCRIPTION
**Pull request by hypeJunction**
Adds a gatekeeper service integrated with the routing system.
Deprecates some gatekeeper functions in favor of OOP alternatives.

BREAKING CHANGE
Gatekeeper functions no longer have a return value, instead they
throw an HttpException.

### CODE REVIEW

File: docs/appendix/upgrade-notes/2.x-to-3.0.rst [hunk 1]
```diff
@@ -314,6 +314,7 @@ Deprecated APIs
  * ``elgg_list_entities_from_access_id``: Use ``elgg_list_entities()``
  * ``elgg_batch_delete_callback``
  * ``\Elgg\Project\Paths::sanitize``: Use ``\Elgg\Project\Paths::sanitize()``
+ * ``elgg_group_gatekeeper``: Use ``elgg_entity_gatekeeper()``
```
Comment (jdalsem) [inline docs/appendix/upgrade-notes/2.x-to-3.0.rst]: what if there is no entity... i still would like to test for group membership

Comment (jdalsem) [reply docs/appendix/upgrade-notes/2.x-to-3.0.rst]: ok, so i can use `entity_gate_keeper(elgg_get_pageowner_entity())` to validate if a user can access a page that is only accessible for group_members?

Comment (hypeJunction) [reply docs/appendix/upgrade-notes/2.x-to-3.0.rst]: Yes. Or `ElggGroup::canAccessContent()`

**Comment by jdalsem** [comment]
can you squash the lintfix?

File: engine/classes/Elgg/Gatekeeper.php [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+	/**
+	 * Require that authenticated user has access to entity
+	 *
+	 * @param ElggEntity $entity Entity
+	 * @param ElggUser   $user   User
+	 *
+	 * @return void
+	 * @throws Exception
+	 * @throws HttpException
+	 */
```
Comment (jdalsem) [inline engine/classes/Elgg/Gatekeeper.php]: throws EntityNotFoundException

Comment (hypeJunction) [reply engine/classes/Elgg/Gatekeeper.php]: It's a subclass of HttpException

Comment (jdalsem) [reply engine/classes/Elgg/Gatekeeper.php]: and that is a subclass of Exception...

**Comment by hypeJunction** [comment]
Moved to #11665

**Pull request closed without merging**